### ISSUE TITLE
Typescript types for onInput, onBlur etc..

### ISSUE THREAD

**Issue by sebkolind** [preactjs/preact#2301]
I am trying to pass an Event Handler to `onInput` and `onBlur` but it is complaining about my types for the function itself. The code works, but I cannot find the correct type.

My event handler is:

```
public onInput = ({target}: {target: HTMLInputElement}) => {
    const {value, name, validity} = target;
    this.setInputValue(name, value);
    this.setInputValidity(name, validity);
    this.setInputs();
};
```

While Typescript expects:
```
Type '({ target }: { target: HTMLInputElement; }) => void' is not assignable to type 'GenericEventHandler<HTMLInputElement>'.
  Types of parameters '__0' and 'event' are incompatible.
    Type 'TargetedEvent<HTMLInputElement, Event>' is not assignable to type '{ target: HTMLInputElement; }'.
      Types of property 'target' are incompatible.
        Type 'EventTarget | null' is not assignable to type 'HTMLInputElement'.
          Type 'null' is not assignable to type 'HTMLInputElement'.ts(2322)
```

I am passing it like this:
```
<input onInput={this.onInput} />
```

I've obviously tried to set it to `GenericEventHandler<HTMLInputElement>` but that does not work, since I need the event passed to the function.

I am sure that I'm doing something completely wrong, but I am not sure what. I can see that in React they have `SyntheticEvent` which helps with this.

**Comment by sebastianks** [comment]
... And just as I posted this, I came up with this solution:

```
interface MyStuff {
   onInput: ({target}: {target: EventTarget | null}) => void;
}
```

```
 public onInput = ({target}: {target: EventTarget | null}) => {
    const {value, name, validity} = target as HTMLInputElement;
    this.setInputValue(name, value);
    this.setInputValidity(name, validity);
    this.setInputs();
};
```

And no the errors are gone when I use it on my input element. But, is this the correct way of doing this, or is it just a hack?

**Comment by ForsakenHarmony** [comment]
which version of preact are you using?

**Comment by sebastianks** [comment]
@ForsakenHarmony Sorry for not providing you with that information. Gonna post the relevant packages here:

```
"devDependencies": {
    "css-loader": "^3.4.2",
    "eslint": "^6.0.1",
    "eslint-config-synacor": "^3.0.4",
    "if-env": "^1.0.0",
    "preact-cli": "^3.0.0-rc.6",
    "ts-loader": "^6.2.1",
    "tslint-config-prettier": "^1.18.0",
    "tslint-eslint-rules": "^5.4.0",
    "tslint-react": "^4.2.0",
    "typescript": "^3.7.5",
    "typings-for-css-modules-loader": "^1.7.0",
    "webpack": "^4.41.5"
},
"dependencies": {
    "preact": "^10.1.0",
    "preact-render-to-string": "^5.1.4",
    "tslint": "^6.0.0"
}
```

**Comment by 38elements** [comment]
This works.

```ts
import { h, render, Component, JSX } from 'preact'


class Foo extends Component {
  onInput({currentTarget}: JSX.TargetedEvent<HTMLInputElement, Event>) {
    console.log(currentTarget.value)
  }
  render() {
    return <input onInput={this.onInput} />
  }
}

render(<Foo />, document.body)
```

**Comment by zachasme** [comment]
Is it possible to get this to work without JSX as well?

The following

```ts
import { h, render, Component, JSX } from 'preact';

class Foo extends Component {
  onInput({ currentTarget }: JSX.TargetedEvent<HTMLInputElement, Event>) {
    console.log(currentTarget.value);
  }
  render() {
    //return <input onInput={this.onInput} />;
    return h('input', { onInput: this.onInput });
  }
}

render(<Foo />, document.body);
```

complains that

```
No overload matches this call.
  Overload 1 of 2, '(type: string, props: (HTMLAttributes<EventTarget> & SVGAttributes<SVGElement> & Record<string, any>) | null, ...children: ComponentChildren[]): VNode<...>', gave the following error.
    Type '({ currentTarget }: TargetedEvent<HTMLInputElement, Event>) => void' is not assignable to type 'GenericEventHandler<EventTarget> & GenericEventHandler<SVGElement>'.
      Type '({ currentTarget }: TargetedEvent<HTMLInputElement, Event>) => void' is not assignable to type 'GenericEventHandler<EventTarget>'.
        Types of parameters '__0' and 'event' are incompatible.
          Type 'TargetedEvent<EventTarget, Event>' is not assignable to type 'TargetedEvent<HTMLInputElement, Event>'.
            Type 'TargetedEvent<EventTarget, Event>' is not assignable to type '{ readonly currentTarget: HTMLInputElement; }'.
              Types of property 'currentTarget' are incompatible.
                Type 'EventTarget' is missing the following properties from type 'HTMLInputElement': accept, align, alt, autocomplete, and 283 more.
  Overload 2 of 2, '(type: ComponentType<{ onInput: ({ currentTarget }: TargetedEvent<HTMLInputElement, Event>) => void; }>, props: (Attributes & { ...; }) | null, ...children: ComponentChildren[]): VNode<...>', gave the following error.
    Argument of type '"input"' is not assignable to parameter of type 'ComponentType<{ onInput: ({ currentTarget }: TargetedEvent<HTMLInputElement, Event>) => void; }>'.ts(2769)
```

**Comment by drackp2m** [comment]
I currently do something like the following:

```
import { h, render, Component } from 'preact';

class Foo extends Component {
  onInput(e: Event) {
    if (e.target instanceof HTMLInputElement) {
      console.log(e.target.value);
    }
  }

  render() {
    return h('input', { onInput: this.onInput });
  }
}

render(<Foo />, document.body);
```


As explained in this section of the documentation: https://preactjs.com/guide/v10/typescript#typing-events

**Comment by m1aw** [comment]
I believe there's a problem with the type `h.JSX.TargetedEvent<>`; 
If I type something like this:
```js
    onInput?: (event: h.JSX.TargetedEvent<HTMLInputElement>) => void;
```
or
```js
    onInput?: GenericEventHandler<HTMLInputElement>;
```
When I pass a function I always get
```
TS2339: Property 'value' does not exist on type 'EventTarget'.
```
```js
   onInput: e => this.handler(e.target.value),
```
Because target is `EventTarget` instead of `HTMLInputElement`.

**Comment by rauschma** [comment]
@m1aw Try `.currentTarget`.

**Comment by marvinhagemeister** [comment]
This is fixed in 10.19.3 . There the type for `event.currentTarget` matches the element the event handler was set on.
